### DISCUSSION TITLE
레벨3 팀 프로젝트 최종 회고

### DISCUSSION THREAD

**Discussion by aak2075** [woowacourse-teams/2023-haru-study#447] [daily_meeting]
## 1. 팀 문화 돌아보기
- 우리에게 의미 있었던 팀 문화를 뽑아보고, 문화가 만들어진 과정과 더불어 어떻게 잘 유지할 수 있었는지 다함께 정리해본다.
- 팀 프로젝트가 기존에 원했던 방향대로 잘 흘러갔는지 각자의 의견 및 감정을 공유하고, 레벨4에 어떻게 유지/개선할 수 있을지 함께 이야기를 나눠본다.
## 2. 나 돌아보기
- 팀 프로젝트를 진행하면서 각자에게 인상 깊었던 사건을 뽑고, 해당 순간에 대한 본인의 감정을 공유한다.
- 좋았던 점(배운 점, 고마운 일), 아쉬웠던 점을 인당 최소 1개 이상 작성한다.
- 레벨3에서 내가 어떤 역할을 했는지 돌아보면서, 레벨4때 어떤 역할을 할 수 있을지 고민해본다.
## 3. 서비스 제작 후기 (팀 단위로 1개만 작성)
- 다함께 힘을 합쳐 하나의 서비스를 만들어 본 후기를 5-6줄 정도로 작성해본다. (기술블로그에 서비스 소개와 함께 올라갈지도^^!)
- 각자 제작 후기를 이야기해보고, 의견을 종합해 하나의 글만 올려주세요.

**Comment by MoonJeWoong** [comment]
# 1. 팀 문화 돌아보기
- 여러 방면으로 좋은 팀문화를 가져가고 싶어서 선정한 액션 아이템은 굉장히 많았는데 일정이 많이 바빠졌던 후반부 스프린트에서는 거의 잘 지키지 못했던 것 같아 많이 아쉽다.
- 그래도 가장 의미있었던 부분은 체크인, 체크아웃 점수를 적으며 오늘 해야할 일, 완료한 일에 대해 공유하는 시간은 빼먹지 않았던 것이었다. 덕분에 바쁜 일정 속에서 각자 하는 일에 대해 지속적인 동기화를 이룰 수 있었다.
- 몹, 페어 프로그래밍으로 기능 구현을 진행한 덕분에 코드 리뷰에 대한 부담을 상대적으로 덜면서 팀원 모두가 코드에 대한 이해를 최대한 많이 가져갈 수 있었던 것 같다.
- 현재 선정한 액션 아이템에서 지키지 못했던 것들은 바빠서 까먹거나 넘기게 되는 경우가 대부분이었는데 매일 혹은 일정 주기로 해당 액션 아이템을 잘 지킬수 있도록 액션 당번을 정해서 팀이 잘 지킬 수 있도록 리마인더 역할을 해주면 좋을 것 같다.

# 2. 나 돌아보기
- 3차 스프린트 발표를 준비하면서 그때 그때 문서화해둔 자료들이 미래에 정말 큰 도움이 된다는 것을 직접적으로 경험할 수 있었다. 발표해야할 내용이나 지켜야할 요구사항이 확 늘어나서 부담이 많이 됐었는데 그 당시 고민들을 잘 기록해둔 것들을 보면서 준비하니 맨땅에서 다시 기억하면서 하는 것보다 훨씬 수월할 수 있었던 것 같다.
- 팀원들이 기술적으로 어떻게 고민하는지 볼 수 있었고 함께 고민할 수 있었어서 정말 많이 배웠던 것 같다.
- 그리고 모르는 부분들을 팀원들에게 물어볼 때마다 항상 친절하게 답변해줬다. 모르는 것을 편하게 얘기할 수 있는 분위기를 함께 만들어 줬음에 감사한다.
- 아쉬웠던 점은 아직 스스로 기술적인 고민을 깊게 해보는 습관이 아직 숙련되지 않아 스스로 고민해서 팀에게 먼저 의견을 제시해보지 못한 것이다.
- 레벨3에서 나의 역할은 팀원들의 의견을 수렴, 종합하고 잘 정리하는 것 + 서로 입장차이로 의사소통이 버거워질 때 이를 잘 해소하는 것이었다고 생각한다.
- 레벨4에서는 소프트 스킬적으로 팀원들의 의견을 잘 조율하고 정리하는 것 외에도 기술적으로 팀에게 이바지하는 역할을 잘 해내고 싶다.

**Comment by yeopto** [comment]
### 1. 팀 문화 돌아보기
- 의미 있었던 팀 문화
  - 데일리 미팅 문화가 의미 있었다. 자연스럽게 아침마다 체크인 점수 및 일정을 공유하고 해야할 일을 정하고 하루를 시작할 수 있어서 효율적이었던 것 같다. 
- 팀 프로젝트에 대한 의견 및 감정
  - 결과적으로는 프로젝트가 잘 진행되었다고 생각한다. 확실히 처음보다 회의시간도 많이 줄어들었다. 하지만 아쉬운 점은 일정추산이 잘 안되었다는 것인데, 레벨 3의 경험 바탕으로 레벨 4때는 일정 추산을 조금 더 현실적으로 잡을 수 있지 않을까 생각한다. 전반적으로 우여곡절도 있었어서 힘들었지만 잘 해결해나간 것 같다.

### 2. 나 돌아보기
- 인상 깊었던 사건
  - 프론트엔드 팀원 룩소가 코로나라 데모 데이 전날에 노아랑 둘이 새벽까지 고치고 배포했던 경험이 인상 깊었다. 이거 왜 안되지를 반복하며 저녁부터 새벽까지 답답했었는데 해결이 되었을 때 카타르시즘을 느꼈다. 
- 좋았던 점(배운 점, 고마운 일)
  - 팀원 전체가 궁금한 점이 있으면 잘 알려준 부분에 대해 감사하다. 그만큼 성장했다.
  - 가끔 직설적으로 말할 때가 있는데, 뒷끝이 없어서 정말 고맙다.
- 아쉬운 점
  - 답답할 때 가끔 직설적으로 말해서 팀원들에게 미안하다.
  - 시간에 쫓겨 하고 싶은 것들을 못한 것들이 있는데, 이 부분들이 좀 아쉽다.
- 레벨 3에서 나의 역할 및 레벨 4때의 역할
  - 레벨 3에서 나의 역할은 분위기 메이커였다고 생각한다. 레벨 4때는 프론트와 백엔드 사이의 징검다리 역할을 잘 하여 프로젝트가 잘 흘러갈 수 있도록 팀에 이바지하고 싶다.

**Comment by jaehee329** [comment]
### 1. 팀 문화 돌아보기
- 처음에 팀 문화를 이것저것 정했고 다 기억하고 지키진 못했지만 그래도 주요한 것들은 자연스럽게 한 것 같아요. 우리의 편의보다 사용자의 편의를 우선하는 방향으로, 그리고 결과에 대해 한 명의 책임으로 돌리지 않고 함께 책임지는 것 등... 그리고 처음에 정하진 않았지만 자연스럽게 다들 6시 이후까지 남아있는 것... 이건 그런 성향의 사람들이 운 좋게 모여서 이렇게 된 것 같아요. 무조건 좋은 건 당연히 아니지만 그래도 같이 열심히 하는구나 하는 느낌을 받았습니다.
- 기간이 기니까 가끔 지치는 사람도 생기고, 또 지치는 순간들이 달라서 그렇지 않은 사람과의 속도가 차이가 나다보니 어려운 순간들이 간간히 나왔다고 느꼈어요. 레벨 4도 비슷한 기간이지만 중간 중간 다른 미션들도 진행해야 하고 개인 공부들도 다들 해야 하니 프로젝트에 얼마나 리소스를 쓸 수 있을까 하는 걱정이 있어요. 모두가 언제나 열정이 넘치긴 어려우니 적당한 수준으로, 함께 나아갈 수 있도록 같이 노력하면 좋겠습니다.

### 2. 나 돌아보기
- 제일 기억나는건 선릉 트랙룸서 저녁먹고 계속 기획회의한거... 힘든데 그냥 또 대충하자 할 수도 없고... 근데 또 다들 힘들어하고..ㅋㅋㅋ
- 팀적으로 잘 어우러지기에 많이 부족했고, 많이 배웠지만 여전히 어려운 순간들이 있었습니다. 내 열정이 넘쳐서 답답해하기도 했고 가끔은 쉬엄쉬엄 하고싶어서 작업을 천천히 하기도 했고요. 팀으로 함께할 때 감정적인 사람이 되면 안 되겠다 싶어 최근에는 더욱 캠퍼스에선 밋밋한 사람이 되려고 했는데 이게 더 나은 방법인지는 아직도 모르겠네요 ㅎㅎㅎ 아직 갈 길이 멀지만 여러모로 노력해볼게요. 
- 항상 이것저것 어디선가 배워와서 새로 공부할 거리들을 던져주는 백엔드 팀원들 덕분에 느려지는 성장 곡선에 가속을 한 번씩 줄 수 있었습니다. 고마워요. 앞으로도 새로운 것들 같이 배워나갔으면 좋겠어요. 특별한 역할을 맡는게 아닌 4에도 지금처럼 같이 나아갔으면 합니다.

**Comment by aak2075** [comment]
### 1. 팀 문화 돌아보기

- 솔직하게 말하면 너무 바쁜 일정 탓에 대부분의 팀 문화는 뒷전으로 미루어졌다. 그나마 쉬는 시간이나 식사 시간을 지키는 문화는 정확하게는 아니지만 잘 지켜졌던 것 같다. 문화를 만들 때에는 정말 많은 시간을 들여서 많은 규칙들을 만들었지만, 실제로 다 지키는 것은 불가능할 수밖에 없다고 생각한다. 정말 필수적인 것들만 추려내서 그것들 만큼은 잘 지키는 것이 좋을 것 같다.
- 팀 프로젝트의 방향은 잘 흘러갔으나 부족한 기능들이 있었다. 백엔드의 리팩토링 영향이 프론트에 크게 미쳐 일정이 밀리는 상황도 발생했었고 백엔드에서도 리팩토링에 너무 많은 시간을 쏟은 것 같다. 리팩토링을 자주하는 것은 좋지만 리팩토링도 몹으로 진행하다 보니 일정이 밀릴 수밖에 없었는데 리팩토링 방향만 함께 정하고 각자 일을 나눠서 진행한다면 리팩토링도, 새로운 기능의 추가나 변경도 대응할 수 있지 않을까 생각한다.

### 2. 나 돌아보기

- 좋았던 점 : API명세를 자세하게(예외 메시지와 예외 코드 등) 해달라는 요구사항이 생겼고, 이에 따라 스웨거로 명세를 하다보니 컨트롤러가 복잡해지고 중복이 많이 발생했다. 이런 문제를 해결하기 위해 커스텀 어노테이션을 도입하여 명세를 간소화했던 점이 좋았다. 어떤 멋진 기술을 사용한 것은 아니지만 실제로 팀이 겪은 문제를 코드로 해결했던 과정이 뿌듯했다.
- 아쉬웠던 점 : 프론트엔드를 위해 API명세를 할 때 프론트엔드에서 어떤 정보가 필요한지 파악하지 않았다. 명세를 해달라고 하니까 필요하다고 생각한 정보만 명세를 하게 되었고, 명세가 명확하지 않아 프론트엔드도 개발이 밀리고 백엔드도 명세를 다시 하는 등 비효율적으로 진행됐다. 다음부터는 그냥 명세를 하는것이 아니라 어떤 정보를 어떻게 전달할 것인지 먼저 고민해야겠다는 생각을 했다.
- 우리 팀은 레벨3에서 특정 사람이 특정한 역할만 맡았다기 보다 필요에 따라 역할을 수행했다. 레벨4에서도 어떤 특별한 역할을 맡는다기 보다 우리 팀이 함께 성장할 수 있는 발판이 되고싶다.

**Comment by woo-jk** [comment]
## 1. 팀 문화 돌아보기
데일리 미팅(체크인, 체크아웃)
- 그날 하루의 시작과 끝 시점의 기분을 팀원들에게 공유할 수 있어 좋았다. 

데모데이 회식
- 3차를 제외하고 데모데이마다 팀 회식을 즐겼다. 서로 친밀도도 쌓을 수 있는 기회였고, 프로젝트간 하고 싶었던 말들을 풀어낼 수 있는 기회여서 좋았다.

회의 공간 옮기기
- 서비스 기획 기간에 자주 했던 방법인데, 회의가 좀 길어진다 싶으면 회의 장소를 강의장이 아닌 무중력 광장 같은 곳으로 옮겨서 진행했다. 장소를 옮겨 회의를 하니 분위기도 환기되고 더 자유로운 분위기에서 회의를 할 수 있어 좋았다.

## 2. 나 돌아보기
가장 인상 깊었던 순간: 기획을 틀어버렸을 때
- 우리 팀 모두 '실제 사용자가 있는 의미있는 서비스'를 만드는 것에 대해 동감했다. 그렇기 때문에 의미있는 서비스를 위해 기획 단계에서 회의를 정말 정말 많이했다. 하지만 처음 기획했던 플랫폼 서비스 기획에 대해 사용자에게 정말로 의미가 있을지 계속해서 의문이 생겼고, 결국 코치분의 조언에 따라 기획을 거의 바꿔버렸다. 2주간 매일 회의하여 얻어낸 기획을 틀어버리는 것은 쉽지 않은 결정이었을텐데 다들 동의 해줘서 고마웠다. 바뀐 기획은 팀원 모두가 마음에 들어하는 것 같고, 서비스 기획이 좋으니 프로젝트에 더욱 애착이 간다.

좋았던 점
- 레벨3 기간동안 정말 찐한 경험을 했다. 더 좋은 기획, 더 좋은 코드를 만들어가기 위해 끊임없이 소통했다. 상대방이 이해를 못하면 설명을 했고, 내가 이해가 안되면 질문을 했다. 소통하는 능력이 많이 부족하다고 생각했는데 레벨3 동안 이 부분에 많이 성장했다.

아쉬운 점
- 열띤 토론을 하고 집에 와서 생각해보면 내가 내 의견을 너무 강하게 밀어붙였나 하고 생각이 들 때가 가끔씩 있었다. 소통을 하면서 중간중간 나를 돌아보고, 팀원들을 돌아보는 시간을 가져야겠다.

내 역할
- 팀원들 각자가 특정한 역할을 했다기보다, 다같이 팀을 잘 이끌어 나간 것 같다.

**Comment by nlom0218** [comment]
# 1. 팀 문화 돌아보기

- 데일리미팅(체크인, 체크아웃): 일과의 시작을 데일리미팅의 체크인으로 시작하고 일과의 끝을 데일리미팅의 체크아웃으로 마무리하였습니다. 이 시간을 통해 오늘 하루 동안의 목표는 무엇이며 앞으로 해야 할 작업이 어떤 것인지 있는지 확인할 수 있어 프로젝트의 방향을 잡는데 큰 도움이 되었습니다. 또한 팀원들의 그날의 감정도 알 수 있어, 함께 공감하는 시간도 가질 수 있었습니다.

- 엄지로 칭찬하기: 프로젝트 초반 기획에 관한 회의를 자주 할 때, 회의 시간이 길어져 서로가 지치는 시간이 있었습니다. 이때 분위기와 텐션을 높이고자 좋은 의견을 제시한 크루에게 즉각적으로 엄치를 세우며 칭찬하는 액션 아이템을 실행하였습니다. 정말 어렵지 않는 하나의 액션 아이템으로 인해 큰 에너지가 소모되는 장시간의 회의가 조금이나마 활력을 찾게 되었습니다. 또한 부수효과로 빠르게 의견을 하나로 모을 수 있었고 다 함께 의견에 대한 동기화도 이전과 비교하여 더욱 빠르게 할 수 있었습니다.

- 혼자서 merge하지 않기: 함께 작업한 PR은 별다른 코드리뷰 없이 바로 merge합니다. 하지만 각자가 작업한 내용에 대한 PR은 반드시 코드 리뷰를 남겨야만 merge가 가능합니다. 이러한 문화로 직접 작성한 코드가 아니더라도 코드의 흐름을 알 수 있어 리팩터링이 필요한 상황이라면 누군든지 이슈를 생성해서 작업이  가능했습니다. 또한 놓쳤던 컨벤션과 코드 품질에 대해서도 팀원들과 함께 이야기를 나눌 수 있었습니다.

- 팀 프로젝트가 기존에 원했던 방향대로 잘 흘러갔는지?: 첫 1,2주 때를 되돌아 생각해보면 서비스의 방향이 명확하지 않아 많은 시간을 회의에 투자를 했어도 `과연 우리 서비스가 매력적인가?`라는 의문을 품었습니다. 하지만 3주차 때 다시 하루스터디만의 핵심 가치를 전달하기 위해 서로가 고민을 하였고 지금은 어디에서도 본 적 없는 매력적인 서비스가 만들어졌다고 생각하여 지금은 뿌듯합니다. 하지만 아직 원하는 기능 구현이 모두 완성되었다고 생각하지 않습니다. 때문에  레벨4를 통해 추가적인 기능 구현과 더불어 고도화를 진행하면 더욱 완성도 높은 프로젝트가 될 수 있으리라 생각합니다. 레벨4 땐, 미션도 많이 있지만 다들 힘내봐요:) 다시 돌아오지 않잖아요 😁

# 2. 나 돌아보기

- 가장 인상깊었던 순간: 이 순간은 아무래도 첫 배포에 성공했을 때입니다. 로컬에서 msw로만 돌아가던 서비스가 실제로 url를 가지고 다른 크루들에게 공유할 수 있었던 순간이 정말 뜻 깊었습니다. 물론 배포 과정에서 오류가 나타나 이를 해결하는데 시간이 걸렸지만 결과적으론 성공적으로 배포를 하게 되었고 지금도 가끔씩 하루스터디를 직접 사용하면서 학습의 효율을 높이고 있습니다.

- 좋았던 점과 아쉬웠던 점: 문제를 공유해주고 해결을 함께 해주려는 팀원들에게 항상 고마움을 느끼고 있어요. 그리고 '이정도면 충분하겠지?'라고 생각하는 순간에 팀원들의 꼼꼼함이 저를 지탱해주었습니다. 이와 같은 도움을 저도 레벨4 땐 팀원들에게 나누어 주고 싶습니다. 아쉬웠던 점은 각자의 기술 공유가 잘 정리되지 않았다는 것입니다. 기술 블로그가 존재하는데 여기에 서로의 기술글을 작성하고 이를 시간을 내어 함께 공부한다면 레벨4 땐 지금보다 함께 성장할 수 있을 것이라 생각합니다. 또한 백엔드+프론트엔드 협업도 해보니 재밌었는데,,, 시간이 된다면 몇 번 더 진행하고 싶습니다.

- 내 역할: 항상 긍정적인 모습을 보이려고 노력을 하는데, 팀원들이 어떻게 받아드렸는지는 잘 모르겠어요. 앞으로도 긍정적인 마음을 잃지 않고 팀원들에게 심적으로 큰 도움을 주고 싶어요.

**Comment by woosung1223** [comment]
## 1. 팀 문화 돌아보기
- 팀 문화에 대해서

좋은 팀 문화를 정착시키려 나름 의미있는 시도를 많이 했던 것 같다. KPT 회고, 데일리 미팅부터 시작해 회의시간 정하고 시작하기, 좋은 의견을 낸 사람에게는 따봉 날려주기 등... 매번 데일리 미팅 때 이런 팀 문화에 대해 이야기를 많이 나눴기 때문에 초기에는 잘 유지할 수 있었던 것 같다.

하지만 개발을 시작하고 여유가 없어지자 팀 문화에 대해서 신경을 많이 쓰지 못했던 것 같다. 특히 데모데이가 가까워지는 날에는, 우리가 만든 규칙들을 '어쩔 수 없으니까' 라는 명목 하에 스스로 깨는 경우가 많았다. 그래서 아쉬움이 많이 남는다. 바쁜 일정 속에서 팀 문화 또한 trade-off의 대상이라고는 생각하긴 하지만 말이다.

- 팀 프로젝트에 대해서

레벨 3에서의 팀 프로젝트는 나름 괜찮았다고 생각한다. 적어도 목표한 기능은 어느정도 구현할 수 있었고, 실제로 프로덕트를 사용할 수 있을 정도는 된다. 그리고 하루스터디 서비스를 사용하는 실사용자도 생겨서 기분이 좋다.

하지만 중간 중간에 어려움은 많았던 것 같다. 모두가 협업 프로젝트는 처음이고, 이렇다할 가이드라인을 가지고 있지 않았기 때문에 헤매는 경우가 많았다. 

RESTful한 API 설계를 하지 못해 중간에 갈아엎기도 했고, 기능 구현이 급한 상황에서 리팩토링에 너무 많은 시간을 소요하기도 했다. 또한 기획이나 기술적인 측면에서 많은 시간을 토론하는 바람에, 팀원들 모두 지치는 경우도 있었다.


## 2. 나 돌아보기
- 가장 인상깊었던 순간

가장 인상깊었던 순간은 아마 기획 회의를 주구장창 하던 때가 아니었나 싶다. 한창 기획 회의를 했을 때에는 저녁도 안먹고 8시까지 토론하던 때도 있었다. 다들 진심으로 좋은 프로덕트를 만들고 싶었기 때문이다. 긴 회의에 지치기는 하면서도 뭔가 '이사람들이라면 뭔가는 만들겠다' 라는 감정을 느꼈었다.

- 좋았던 점과 아쉬웠던 점

좋았던 점은 기획, 협업에 대해 많이 배웠다는 것이다. 우리 하루스터디 팀은 개인보다는 공동체라는 느낌이 강했다. 회의에 대한 거부감도 없고, 의사소통에 대한 두려움도 없어서 토론을 정말 많이 했다. 이 과정에서 찐한 협업을 경험할 수 있었다.

아쉬웠던 점은 기술적으로는 많이 성장하지 못했다는 것이다. 학습보다는 '프로덕트 만들기'에 집중하다보니 스스로 가져가는게 그렇게 많지는 않았던 것 같다. 

- 레벨 3에서의 나의 역할

프로젝트 초반에는 팀 문화 정착에 기여를 어느정도 했다고 생각한다. KPT 회고부터 Action Item 선정하는 것까지, 구글링을 통해 좋은 문화에 대해 찾아보고 알아내서 제시한 것들이다.

프로젝트 중에는 나만의 특별한 역할은 따로 없었던 것 같고, 팀원들 모두가 특별하게 잘 해주었던 것 같다. 앞으로도 이대로만 가면 어떨까 싶다.
